In [1]:
import pandas as pd
import bamboolib as bam
cars_data = pd.read_csv("cars.csv")


cars_data['name'] = cars_data['name'].astype('string')

cars_data = cars_data.rename(columns={'cylenders': 'cylinders'})
cars_data

name  sports_car    suv  wagon  minivan  \
0                 Chevrolet Aveo 4dr       False  False  False    False   
1        Chevrolet Aveo LS 4dr hatch       False  False  False    False   
2             Chevrolet Cavalier 2dr       False  False  False    False   
3             Chevrolet Cavalier 4dr       False  False  False    False   
4          Chevrolet Cavalier LS 2dr       False  False  False    False   
..                               ...         ...    ...    ...      ...   
423         Nissan Titan King Cab XE       False  False  False    False   
424                      Subaru Baja       False  False  False    False   
425                    Toyota Tacoma       False  False  False    False   
426     Toyota Tundra Regular Cab V6       False  False  False    False   
427  Toyota Tundra Access Cab V6 SR5       False  False  False    False   

     pickup  all_wheel  rear_wheel  Price  Dealer_Cost  Engine_size  \
0     False      False       False  11690        10965          1.6   
1     False      False       False  12585        11802          1.6   
2     False      False       False  14610        13697          2.2   
3     False      False       False  14810        13884          2.2   
4     False      False       False  16385        15357          2.2   
..      ...        ...         ...    ...          ...          ...   
423    True       True       False  26650        24926          5.6   
424    True       True       False  24520        22304          2.5   
425    True      False        True  12800        11879          2.4   
426    True      False        True  16495        14978          3.4   
427    True       True       False  25935        23520          3.4   

     cylinders  horsepower  city_miles_per_galloon  highway_miles_per_Gallon  \
0            4         103                    28.0                      34.0   
1            4         103                    28.0                      34.0   
2            4         140                    26.0                      37.0   
3            4         140                    26.0                      37.0   
4            4         140                    26.0                      37.0   
..         ...         ...                     ...                       ...   
423          8         305                    14.0                      18.0   
424          4         165                    21.0                      28.0   
425          4         142                    22.0                      27.0   
426          6         190                    16.0                      20.0   
427          6         190                    14.0                      17.0   

     weight  base_wheeel  length  width  
0    2370.0         98.0   167.0   66.0  
1    2348.0         98.0   153.0   66.0  
2    2617.0        104.0   183.0   69.0  
3    2676.0        104.0   183.0   68.0  
4    2617.0        104.0   183.0   69.0  
..      ...          ...     ...    ...  
423  5287.0        140.0     NaN    NaN  
424  3485.0        104.0     NaN    NaN  
425  2750.0        103.0     NaN    NaN  
426  3925.0        128.0     NaN    NaN  
427  4435.0        128.0     NaN    NaN  

[428 rows x 19 columns]

### 1- How many Rows are in the cars dataset?
    cars_data_rows_num = cars_data.shape[0]
'428 rows'

### 2- How many Columns are in the car's data set?
    cars_data_cols_num = cars_data.shape[1]; display(f"{cars_data_cols_num} columns")
    '19 columns'

### 3- How many unique numbers of cylinders we have in the cars dataset?
    num_of_unique_values_of_cylinders = cars_data["cylinders"].nunique()

        8

### 4- what is the average horsepower of cars? 
    avg_horsepower = cars_data["horsepower"].mean()

        215.8855140186916

### 5- what is the maximum horsepower?
    max_horsepower = cars_data['horsepower'].max()

        
        500
### 6- what is the most expensive car?
    max_value = cars_data['price'].max()

    max_rows = cars_data.loc[cars_data['price']==max_value]
    most_expensive_car = max_rows.iloc[0,0]
        
        'Porsche 911 GT2 2dr'

### 7- change the name of the column "name" to "car name"
    cars_data = cars_data.rename(columns={'name': 'car name'})


In [2]:
### 1- How many Rows are in the cars dataset?

cars_data_rows_num = cars_data.shape[0]; display(f"{cars_data_rows_num} rows")

'428 rows'

In [3]:
### 2- How many Columns are in the car's data set?
cars_data_cols_num = cars_data.shape[1]; display(f"{cars_data_cols_num} columns")

'19 columns'

In [4]:
### 3- How many unique numbers of cylinders we have in the cars dataset?

num_of_unique_values_of_cylinders = cars_data["cylinders"].nunique()
num_of_unique_values_of_cylinders

8

In [5]:
### 4- what is the average horsepower of cars? 

avg_horsepower = cars_data["horsepower"].mean()
avg_horsepower

215.8855140186916

In [6]:
### 5- what is the maximum horsepower?

max_horsepower = cars_data['horsepower'].max()
max_horsepower

500

In [7]:
### 6- what is the most expensive car?
    
max_value = cars_data['Price'].max()

max_rows = cars_data.loc[cars_data['Price']==max_value]
most_expensive_car = max_rows.iloc[0,0]

most_expensive_car

'Porsche 911 GT2 2dr'

In [8]:
### 7- change the name of the column "name" to "car name"

cars_data = cars_data.rename(columns={'name': 'car name'})
cars_data

car name  sports_car    suv  wagon  minivan  \
0                 Chevrolet Aveo 4dr       False  False  False    False   
1        Chevrolet Aveo LS 4dr hatch       False  False  False    False   
2             Chevrolet Cavalier 2dr       False  False  False    False   
3             Chevrolet Cavalier 4dr       False  False  False    False   
4          Chevrolet Cavalier LS 2dr       False  False  False    False   
..                               ...         ...    ...    ...      ...   
423         Nissan Titan King Cab XE       False  False  False    False   
424                      Subaru Baja       False  False  False    False   
425                    Toyota Tacoma       False  False  False    False   
426     Toyota Tundra Regular Cab V6       False  False  False    False   
427  Toyota Tundra Access Cab V6 SR5       False  False  False    False   

     pickup  all_wheel  rear_wheel  Price  Dealer_Cost  Engine_size  \
0     False      False       False  11690        10965          1.6   
1     False      False       False  12585        11802          1.6   
2     False      False       False  14610        13697          2.2   
3     False      False       False  14810        13884          2.2   
4     False      False       False  16385        15357          2.2   
..      ...        ...         ...    ...          ...          ...   
423    True       True       False  26650        24926          5.6   
424    True       True       False  24520        22304          2.5   
425    True      False        True  12800        11879          2.4   
426    True      False        True  16495        14978          3.4   
427    True       True       False  25935        23520          3.4   

     cylinders  horsepower  city_miles_per_galloon  highway_miles_per_Gallon  \
0            4         103                    28.0                      34.0   
1            4         103                    28.0                      34.0   
2            4         140                    26.0                      37.0   
3            4         140                    26.0                      37.0   
4            4         140                    26.0                      37.0   
..         ...         ...                     ...                       ...   
423          8         305                    14.0                      18.0   
424          4         165                    21.0                      28.0   
425          4         142                    22.0                      27.0   
426          6         190                    16.0                      20.0   
427          6         190                    14.0                      17.0   

     weight  base_wheeel  length  width  
0    2370.0         98.0   167.0   66.0  
1    2348.0         98.0   153.0   66.0  
2    2617.0        104.0   183.0   69.0  
3    2676.0        104.0   183.0   68.0  
4    2617.0        104.0   183.0   69.0  
..      ...          ...     ...    ...  
423  5287.0        140.0     NaN    NaN  
424  3485.0        104.0     NaN    NaN  
425  2750.0        103.0     NaN    NaN  
426  3925.0        128.0     NaN    NaN  
427  4435.0        128.0     NaN    NaN  

[428 rows x 19 columns]

In [9]:
### 8- make a subset of the data that has the car name, the price and name the new sub-setted data frame  car pricing.
car_pricing = cars_data[['car name', 'Price']]
car_pricing_12545 = car_pricing.loc[car_pricing['car name'].isin(['Chevrolet Aveo 4dr'])]
car_pricing_12545.iloc[0,1]

11690

In [10]:
### 9- create a function called pricing category that returns "Budget Car " if the cars are less than 20,000 USD," Suitable Car " 
### is the car is more than 20,000 and less than 35 000 and finally an expensive car for cars more than 35000.

def pricing_category(car_name):
    car_pricing = cars_data[['car name', 'Price']]
    try:
        car_name_df = car_pricing.loc[car_pricing['car name'].isin([car_name])]
        car_price = car_name_df.iloc[0,1]
        
        if car_price < 20000:
            return "Budget Car"
        elif car_price > 20000 and car_price < 35000:
            return "Suitable Car"
        elif car_price > 35000:
            return "Expensive car"
        
    except (NameError, TypeError, ZeroDivisionError):
        print("Please make sure to write the name of the care correctly!")

the_car_name = input("Please Enter the name of the car, which you need to know its category: ")
print(pricing_category(the_car_name))
            

Suitable Car
Suitable Car


In [11]:
### 10- create a column named category on the subset using a for loop and pricing category function.

# 1. iterate through pandas dataframe column -> iterate through "car name" column 
for i in range(len(car_pricing)):
    car_pricing.loc[i, "category"] = pricing_category(car_pricing.loc[i, "car name"])
    
car_pricing['category'] = car_pricing['category'].astype('string')
car_pricing

Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Suitable Car
Suitable Car
Budget Car
Suitable Car
Suitable Car
Budget Car
Suitable Car
Budget Car
Suitable Car
Suitable Car
Budget Car
Suitable Car
Budget Car
Suitable Car
Suitable Car
Budget Car
Suitable Car
Budget Car
Budget Car
Suitable Car
Budget Car
Budget Car
Suitable Car
Budget Car
Budget Car
Budget Car
Suitable Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Budget Car
Suitable Car
Suitable Car
Budget Car
Suitable Car
Budget Car
Suitable 

car name  Price      category
0                 Chevrolet Aveo 4dr  11690    Budget Car
1        Chevrolet Aveo LS 4dr hatch  12585    Budget Car
2             Chevrolet Cavalier 2dr  14610    Budget Car
3             Chevrolet Cavalier 4dr  14810    Budget Car
4          Chevrolet Cavalier LS 2dr  16385    Budget Car
..                               ...    ...           ...
423         Nissan Titan King Cab XE  26650  Suitable Car
424                      Subaru Baja  24520  Suitable Car
425                    Toyota Tacoma  12800    Budget Car
426     Toyota Tundra Regular Cab V6  16495    Budget Car
427  Toyota Tundra Access Cab V6 SR5  25935  Suitable Car

[428 rows x 3 columns]

In [12]:
### 11- How many Budget cars, suitable cars, and expensive cars we have?

display((car_pricing["category"]=='Budget Car').sum())
display((car_pricing["category"]=='Suitable Car').sum())
display((car_pricing["category"]=='Expensive car').sum())



98

193

137